Домашнее задание 15 спринта

DaniilKosmachev opened the pull request.

DaniilKosmachev committed:
ДЗ 15, стабильная:
Созданы пакеты
Разделены классы по пакетм, связаны интерфейсами
Запрос в сеть ходит, отдает даже треки

DaniilKosmachev committed:
ДЗ 15, стабильная вер 1:
Разделена ответственность SharedPreferences

DaniilKosmachev committed:
ДЗ 15, стабильная вер 2:
Разделена ответственность SharedPreferences, создается все из Creator.kt

DaniilKosmachev committed:
ДЗ 15, стабильная вер 3:
Разделен MediaPlayer в активити AudioPlayerActivity.kt

DaniilKosmachev committed:
ДЗ 15, стабильная вер 3:
Разделен MediaPlayer в активити AudioPlayerActivity.kt

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/Creator.kt:
```diff
@@ -0,0 +1,17 @@
+import com.example.playlistmaker.domain.api.TracksRepository
+import com.example.playlistmaker.domain.impl.TracksInteractorImpl
+
+object Creator {
```
кажется, что Creator мы добавим только в последующих спринтах. Но ты разобрался хорошо, ничего не поломав. Круто!

DaniilKosmachev replied:
По поводу критического замечания, просто переместить все пакеты из слоя ui в presentation и все?

DaniilKosmachev replied:
^_^

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/data/NetworkClient.kt:
```diff
@@ -0,0 +1,7 @@
+import com.example.playlistmaker.data.dto.Response
+
+interface NetworkClient {
+    fun doRequest(dto: Any): Response
```
думаю стоит дать более конкретное название что за запрос мы делаем, а что за ответ получаем

DaniilKosmachev replied:
Done

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/data/dto/TrackDto.kt:
```diff
@@ -0,0 +1,14 @@
+package com.example.playlistmaker.data.dto
+
+data class TrackDto(
```
если я не ошибаюсь, какие-то из полей опциональные. перепроверь этот момент пожалуйста

DaniilKosmachev replied:
Посмотрел - trackName,artistName,trackTimeMillis-artworkUrl100,collectionName,releaseDate,primaryGenreName,country,prewiewUrl - для PlayerActiviry используются, а traackId - для поиска совпадений в Shared Preferences одинаковых треков (чтобы переместить трек при нажатии на историю на первую позицию в листе)

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/data/local/TrackSearchHistory.kt:
```diff
@@ -1,28 +1,34 @@
 
-    private val sharedPreferencesHistory = sharedPreferences
+class TrackSearchHistory(var sharedPreferences: SharedPreferences) :
+    Serializable, Application() {
```
я думаю, что здесь наследоваться от Application - плохая идея. редко когда какой-то класс в этом нуждается

DaniilKosmachev replied:
Done

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,40 @@
+    private val itunesService = retrofit.create(ITunesApi::class.java)
+
+    override fun doRequest(dto: Any): Response {
+        if (dto is TrackSearchRequest) {
```
как раз таки если мы сдеаем dto не типа Any, не придется делать проверку. Напомню что Any - самый верховный класс, то есть dto может быть вообще чем угодно. Пересмотри пожалуйста этот момент

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/ui/main/MainActivity.kt:
```diff
@@ -1,10 +1,14 @@
+import com.example.playlistmaker.ui.settings.SettingsActivity
+import com.example.playlistmaker.ui.tracks.SearchActivity
 
 class MainActivity : AppCompatActivity(), View.OnClickListener {
```
наследоваться от View.OnClickListener - лишнее, а ставить listener в onCreate не имеет смысла. Просто используй  `binding.libraryWindow.setOnClickListener { }`, а в лямбде (фигурных скобках если что) выполни нужные действия

DaniilKosmachev replied:
Done

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/ui/tracks/SearchActivity.kt:
```diff
@@ -230,7 +213,7 @@ class SearchActivity : AppCompatActivity(), java.util.function.Consumer<List<Tra
     @SuppressLint("NotifyDataSetChanged")
```
эта аннотация на самом деле не играет роли, можно смело удалить

DaniilKosmachev replied:
Done

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/AudioPlayerActivity.kt:
```diff
@@ -1,159 +1,138 @@
         super.onCreate(savedInstanceState)
         binding = ActivityAudioPlayerBinding.inflate(layoutInflater)
         setContentView(binding.root)
+        selectableTrack = if (VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
а вот это уже плохой подход. если пользователь запустит приложение с версией не тирамису, ничего не сработает

DaniilKosmachev replied:
Ниже идет else (где deprecated), где как раз для версии Tiramisu - так советовали наставники

nadzyathedude commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/AudioPlayerActivity.kt:
```diff
@@ -1,159 +1,138 @@
                     PlayerState.PREPARED -> {
+                        binding.playButton.setImageResource(R.drawable.button_play)
                         binding.currentTrackTimeAudioPlayer.text = "00:00"
                     }
```
предлагаю вынести строковое значение в переменную, тем самым избежав хардкода

DaniilKosmachev replied:
Done

DaniilKosmachev committed:
ДЗ 15, стабильная вер 4:
Пакет ui перенесен в пакет presenatation
Найден и исправлен баг - при отсутствии сети крашиться (pool-2-thread-1)
Исправлены некритические замечания
